### ISSUE TITLE
Lint: Check for duplicate `:pre` and `:post` expressions

### ISSUE THREAD

**Issue by NoahTheDuke** [clj-kondo/clj-kondo#1671]
To upvote this issue, give it a thumbs up. See [this list](https://github.com/clj-kondo/clj-kondo/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) for the most upvoted issues.

**Is your feature request related to a problem? Please describe.**
When writing function definitions with `:pre` and/or `:post` assertions, one can sometimes accidentally write the same assertion multiple times.

For example:

```clojure
(defn even-number? [n] {:pre [(number? n) (even? n) (number? n)]} n)
```

**Describe the solution you'd like**
A warning on the second `(number? n)` sexpr, indicating it's a duplicate.

**Describe alternatives you've considered**
There are other linters that check for such things, like Eastwood, but it's nice to have such checks here as well. 

I am willing to write this lint if it's wanted.

**Comment by borkdude** [comment]
Considerations:

1) Can this linter be generalized? E.g. expressions in `and` etc could also be duplicated
2) Could this lead to false positives? Yes, if the expressions contain side effects
3) How important is this: how frequently do mistakes like this happen and how serious is the damage of those mistakes?

The combination of 2 and 3 makes me inclined to say that clj-kondo should (currently) not support this.

**Comment by NoahTheDuke** [comment]
Thoughts on those questions, no solid answers:

1. Yes, the hook I wrote applies to `and`, `or`, and :pre/:post expressions.
2. It could. It gets the hash of each form and then warns if there's duplicates. I don't think we can reasonably check for purity here, as that would probably require tree-walking each form and comparing each function call to a list of known impure functions (or checking the configuration, etc). Feels pretty messy to me. Better to leave it up to the users to determine what is or is not worth validating/exempting.
3. Relatively little importance, like most of these lints. The goal isn't to only implement the biggest and most important linters but to cover as large a space as possible so that users can configure and lint their code base to their standards. (See: Rubocop's quite large number of lints.) For example, my previous lint added checking if the `:keys` vector contained keywords. Keywords in the `:keys` vector has little impact and no danger but for some code bases it's helpful to stay consistent. I don't see why it should be any different now.

**Comment by borkdude** [comment]
> The goal isn't to only implement the biggest and most important linters but to cover as large a space as possible so that users can configure and lint their code base to their standards.

Note that adding code creates a bigger maintenance burden, so I do want to have more than one person requesting that feature. I don't necessarily agree that everything should be covered. Often it's better to wait if linters can be generalized and not add them in a hurry (hammock time), as to not create a zillion very specific linters and a lot of hardly used code to maintain.

**Comment by NoahTheDuke** [comment]
That makes sense, thanks. My apologies if my previous reply came off aggressive. I didn't intend it so when writing but on reread, it feels aggressive.

**Comment by borkdude** [comment]
No problem
